### DISCUSSION TITLE
(Question) About BinaryAST compatibility

### DISCUSSION THREAD

**Discussion by castarco** [denoland/deno#11064] [General]
Back in 2019, Cloudflare announced the BinaryAST proposal ( https://blog.cloudflare.com/binary-ast/ , https://tc39.es/proposal-binary-ast/ ), which in theory could help with cold load times.

Fast-forwarding to now, there haven't been many news on the topic, or at least Google didn't index anything relevant.

Relating to that, I have three questions:
- Would it make sense to "transpile" the typescript checker/compiler code into BinaryAST to improve startup times? or is it already saved as lower level bytecode?
- Would it make sense, or provide any benefit, to allow people running their programs compiled down to BinaryAST?
- Does anyone know how far this kind of feature is? I personally have no idea about whether V8 supports it or not as of today.

**Comment by kitsonk** [comment]
It is a Stage 1 proposal that was last presented in Sept 2018.  With no active advocacy in TC39, I don't think we would consider it.  If it reached Stage 3 and there looked likely an implementation arriving in a browser, that is the point where we would jump on.

Deno already serialises all of its runtime code using the snapshot feature of V8, which in theory would still be faster than the AST byte code. Issue #1877 covers this for users and is likely to be part of of `deno compile` at some point.

Issue #3335 covers caching user byte code automatically, but it has been a very much a back burner issue, because we are pretty convinced it won't make an appreciable difference in startup time.
